### ISSUE TITLE
Index column size too large

### ISSUE THREAD

**Issue by markuman** [nextcloud/server#16448]
### Steps to reproduce
1. Using AWS Aurora 2 (Compatible to MySQL 5.7.12)
2. Update from Nextcloud 15 to Nextcloud 16

### Actual behaviour

In the DB Schema Update, the following sql statement

```sql
CREATE TABLE
	oc_login_flow_v2 (id BIGINT UNSIGNED AUTO_INCREMENT NOT NULL,
	timestamp BIGINT UNSIGNED NOT NULL,
	started SMALLINT UNSIGNED DEFAULT 0 NOT NULL,
	poll_token VARCHAR(255) NOT NULL,
	login_token VARCHAR(255) NOT NULL,
	public_key TEXT NOT NULL,
	private_key TEXT NOT NULL,
	client_name VARCHAR(255) NOT NULL,
	login_name VARCHAR(255) DEFAULT NULL,
	server VARCHAR(255) DEFAULT NULL,
	app_password VARCHAR(1024) DEFAULT NULL,
	UNIQUE INDEX poll_token (poll_token),
	UNIQUE INDEX login_token (login_token),
	INDEX timestamp (timestamp),
	PRIMARY KEY(id)) DEFAULT CHARACTER SET
	utf8mb4 COLLATE utf8mb4_bin ENGINE = InnoDB ROW_FORMAT = compressed;
```

produces the following error

```
SQL-Fehler [1709] [HY000]: Index column size too large. The maximum column size is 767 bytes.
```

In the past (we're running our nextcloud on aws aurora since nextcloud 13) I saw those error often. The trick was to dump the database into a native mysql or mariadb und start the update on that database. when it is finished, I simply dump the updated db and restore it on aurora.
That works fine until today when updating to nextcloud 16.

### Server configuration

* Docker Nextcloud Image 16.0.3
* AWS 5.7.mysql_aurora.2.04.3

**Nextcloud version:** (see Nextcloud admin page)

16.0.3

**Updated from an older Nextcloud/ownCloud or fresh install:**

Update from 15.0.10 to 16.0.3

**Where did you install Nextcloud from:**

Docker

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

The SQL statement work when I removed the `ROW_FORMAT = compressed`.

Are there any disadvantages when the this table is not compressed? Any chance to change this on upstream so that nextcloud is still compatible with aws aurora?

**Comment by markuman** [comment]
I can handle the incompatiblity with proxysql 

```yml
    - name: nextcloud 16 fix
      proxysql_query_rules:
        login_user: "{{ proxy_sql_admin_user }}"
        login_password: "{{ proxy_sql_admin_password }}"
        match_digest: "CREATE.*"
        match_pattern: "ROW_FORMAT=compressed"
        replace_pattern: " "
        destination_hostgroup: "{{ hostgroup_master }}"
        active: 1
        retries: 3
        state: present
        load_to_runtime: True
```
*(_this is ansible syntax_)

AFAIU, the row compression benefits storage usage on cost of cpu usage. 

Any thought about other negative aspects which hit me hard on later updates?

**Comment by blizzz** [comment]
@rullzer apparently login flow v2 related

**Comment by kesselb** [comment]
> Any thought about other negative aspects which hit me hard on later updates?

There is a similar discussion at Moodle: https://tracker.moodle.org/browse/MDL-58931

> Any chance to change this on upstream so that nextcloud is still compatible with aws aurora?

https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/DB/ConnectionFactory.php#L223-L228

Value is hard coded here. We could probably expose it (or defaultTableOptions) to `config.php`.

**Comment by fmmendes** [comment]
> > Any thought about other negative aspects which hit me hard on later updates?
> 
> There is a similar discussion at Moodle: https://tracker.moodle.org/browse/MDL-58931
> 
> > Any chance to change this on upstream so that nextcloud is still compatible with aws aurora?
> 
> https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/DB/ConnectionFactory.php#L223-L228
> 
> Value is hard coded here. We could probably expose it (or defaultTableOptions) to `config.php`.

I've made a test here changing the parameter row_format to dynamic and, after the creation of the tables, I altered the row_format of the tables to compressed, and the command ran successfully but it shows 2 warning(s): 1478 ROW_FORMAT=COMPRESSED is not currently supported 1478 InnoDB: assuming ROW_FORMAT=COMPACT

**Comment by szaimen** [comment]
Is this Issue still valid in NC21.0.2? If not, please close this issue. Thanks! :)

**Comment by markuman** [comment]
We do not use aurora anymore. but as long as the table schema does not change, I bet you'll still get an error when using aurora mysql from aws.

**Comment by blizzz** [comment]
If this is aurora-specific and they forked mysql in an incompatible way, I do not see us fixing it, really. Auroa is not supported, MySQL and MariaDB (amongst others) are.

**Comment by kesselb** [comment]
To reference: https://github.com/nextcloud/server/issues/25436 and https://github.com/nextcloud/server/pull/25474.

**Comment by markuman** [comment]
starting from mariadb 10.6, row format compressed behaves different.

* https://github.com/nextcloud/docker/issues/1492
* https://mariadb.com/kb/en/changes-improvements-in-mariadb-106/
* https://jira.mariadb.org/browse/MDEV-23497


So this aging issue hits mariadb users now - as you've noticed already.

**Comment by markuman** [comment]
This issue can be closed.  

TL;DR: Don't use AWS Aurora as Nextcloud backend.

Fun fact: The issue that breaks AWS Aurora compatibility was introduced with Nextcloud 16 and hits now mariadb 10.6 users: https://jira.mariadb.org/browse/MDEV-23497

proxysql users can remove row format compressed and take care about it: https://github.com/nextcloud/server/issues/16448#issuecomment-512749219  
works still with AWS Aurora and also with mariadb 10.6
